pallets/did/src/lib.rs

```diff
@@ -348,53 +349,43 @@ pub mod pallet {
 		/// - Writes: Did (with K new key agreement keys)
 		/// # </weight>
 		#[pallet::weight(
-      <T as pallet::Config>::WeightInfo::submit_did_create_operation_ed25519_keys(
-        operation.new_key_agreement_keys.len().saturated_into::<u32>(),
-        operation.new_endpoint_url.as_ref().map_or(0u32, |url| url.len().saturated_into::<u32>())
-      )
-      .max(<T as pallet::Config>::WeightInfo::submit_did_create_operation_sr25519_keys(
-        operation.new_key_agreement_keys.len().saturated_into::<u32>(),
-        operation.new_endpoint_url.as_ref().map_or(0u32, |url| url.len().saturated_into::<u32>())
-      ))
-      .max(<T as pallet::Config>::WeightInfo::submit_did_create_operation_ecdsa_keys(
-        operation.new_key_agreement_keys.len().saturated_into::<u32>(),
-        operation.new_endpoint_url.as_ref().map_or(0u32, |url| url.len().saturated_into::<u32>())
-      ))
-    )]
-		pub fn submit_did_create_operation(
-			origin: OriginFor<T>,
-			operation: DidCreationOperation<T>,
-			signature: DidSignature,
-		) -> DispatchResult {
+			<T as pallet::Config>::WeightInfo::create_ed25519_keys(
+				details.new_key_agreement_keys.len().saturated_into::<u32>(),
+				details.new_endpoint_url.as_ref().map_or(0u32, |url| url.len().saturated_into::<u32>())
+			)
+			.max(<T as pallet::Config>::WeightInfo::create_sr25519_keys(
+				details.new_key_agreement_keys.len().saturated_into::<u32>(),
+				details.new_endpoint_url.as_ref().map_or(0u32, |url| url.len().saturated_into::<u32>())
+			))
+			.max(<T as pallet::Config>::WeightInfo::create_ecdsa_keys(
+				details.new_key_agreement_keys.len().saturated_into::<u32>(),
+				details.new_endpoint_url.as_ref().map_or(0u32, |url| url.len().saturated_into::<u32>())
+			))
+		)]
+		pub fn create(origin: OriginFor<T>, details: DidCreationDetails<T>, signature: DidSignature) -> DispatchResult {
 			let sender = ensure_signed(origin)?;
 
+			let did_identifier = details.did.clone();
+
 			// There has to be no other DID with the same identifier already saved on chain,
 			// otherwise generate a DidAlreadyPresent error.
-			ensure!(
-				!<Did<T>>::contains_key(operation.get_did()),
-				<Error<T>>::DidAlreadyPresent
-			);
+			ensure!(!<Did<T>>::contains_key(&did_identifier), <Error<T>>::DidAlreadyPresent);
 
-			let account_did_auth_key = operation
-				.did
-				.verify_and_recover_signature(&operation.encode(), &signature)
+			let account_did_auth_key = did_identifier
+				.verify_and_recover_signature(&details.encode(), &signature)
 				.map_err(<Error<T>>::from)?;
 
-			let did_entry =
-				DidDetails::try_from((operation.clone(), account_did_auth_key)).map_err(<Error<T>>::from)?;
+			let did_entry = DidDetails::try_from((details, account_did_auth_key)).map_err(<Error<T>>::from)?;
```

Comment:
This would look better wouldn't it?

```suggestion
			let did_entry = details.update(account_did_auth_key)?;
```

Also do you really need to map the error? This should happen automatically with the `?`?

Reply:
`details` is what is given in the extrinsic call, so we can't just call `update` on that. We need to create a new instance of `DidDetails` from scratch, and we do that by analysing the information inside the creation details + the authentication key we recover from the signature.

For the second point, it is true that `DidError` automatically maps to `DispatchError`, but `try_from` generates an `InputError`, which we must map to a `DidError` first, which then gets converted to a `DispatchError`.

Reply:
Perhaps we could call `update` from an update details?

Reply:
> `details` is what is given in the extrinsic call, so we can't just call `update` on that. We need to create a new instance of `DidDetails` from scratch, and we do that by analysing the information inside the creation details + the authentication key we recover from the signature.

Yes I agree that `update` doesn't fit 100%. But `update` can return a new instance. I think the convention is to call the fn `updated` than.
`TryFrom` takes one thing and transforms it into another, if possible. I don't feel like it's matching this use case here.  IMO So while update only fits 80%, try from fits like 50%.

> For the second point, it is true that `DidError` automatically maps to `DispatchError`, but `try_from` generates an `InputError`, which we must map to a `DidError` first, which then gets converted to a `DispatchError`.

😂 Ok I see.

Reply:
Maybe something like this? https://github.com/KILTprotocol/mashnet-node/pull/231/commits/d8bddbf9ca78a7d0f3084ee0eac1f781755e782e

Reply:
no need to pass a tuple. 😉 Otherwise it's perfect! 😍 

```
pub fn from_creation_details((details, new_auth_key): (DidCreationDetails<T>, DidVerificationKey)) -> Result<Self, InputError>
```

Reply:
Voilà https://github.com/KILTprotocol/mashnet-node/pull/231/commits/391260f57754d6072b91848e420b4487e229e5b2 🎩